tools/lsp/preview.rs

```diff
@@ -122,50 +135,30 @@ impl ContentCache {
         }
     }
 }
-
-#[derive(Default)]
-struct PreviewState {
-    ui: Option<ui::PreviewUi>,
-    property_range_declarations: Option<ui::PropertyDeclarations>,
-    handle: Rc<RefCell<Option<slint_interpreter::ComponentInstance>>>,
-    document_cache: Rc<RefCell<Option<Rc<common::DocumentCache>>>>,
-    selected: Option<element_selection::ElementSelection>,
-    notify_editor_about_selection_after_update: bool,
-    workspace_edit_sent: bool,
-    known_components: Vec<ComponentInformation>,
-    preview_loading_delay_timer: Option<slint::Timer>,
-    initial_live_data: preview_data::PreviewDataMap,
-    current_live_data: preview_data::PreviewDataMap,
-}
-
-impl PreviewState {
-    fn component_instance(&self) -> Option<ComponentInstance> {
-        self.handle.borrow().as_ref().map(|ci| ci.clone_strong())
-    }
-}
 thread_local! {static PREVIEW_STATE: std::cell::RefCell<PreviewState> = Default::default();}
 
 // Just mark the cache as "read from disk" by setting the version to None.
 // Do not reset the code: We can check once the LSP has re-read it from disk
 // whether we need to refresh the preview or not.
 fn invalidate_contents(url: &lsp_types::Url) {
-    let mut cache = CONTENT_CACHE.get_or_init(Default::default).lock().unwrap();
-
-    if let Some(cache_entry) = cache.source_code.get_mut(url) {
-        cache_entry.version = None;
-    }
+    PREVIEW_STATE.with(|preview_state| {
```

Comment:
Note that there is now a `with_borrow` and `with_borrow_mut` that can be used in new code and would simplify this pattern a bit.
https://doc.rust-lang.org/std/thread/struct.LocalKey.html#method.with_borrow